### ISSUE TITLE
Enable CheckboxField and RadioField the ability to announce error and help text to screen readers

### ISSUE THREAD

**Issue by neilpfeiffer** [cerner/terra-core#1515]
# Issue Description
Add additional ARIA attributes like `aria-labelledby` and `aria-describedby` to helptext and other field elements as needed in all specifically created Form Field Compositions (InputField, TextAreaField, CheckboxField, RadioField, SelectField, etc) for better accessibility. Key area of focus are the Label, Required, Help Text, and Error Text.  

(Note: doesn't apply to Terra-Form-Field since component is abstract without context of which input might be provided, unless using a RenderProps method -- future consideration)

## Issue Type

- [ ] New Feature
- [x] Enhancement
- [ ] Bug
- [ ] Other

## Expected Behavior
Should be similar to existing work already done for form-fields in Legacy Terra et. al. 
Help Text References: 
- https://www.w3.org/WAI/tutorials/forms/instructions/
- https://www.w3.org/TR/WCAG20-TECHS/ARIA1.html
- https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/forms/Basic_form_hints
- https://material.io/develop/web/components/input-controls/text-field/helper-text/

Error Text References:
- https://www.w3.org/WAI/GL/wiki/Using_ARIA_role_of_alert_for_Error_Feedback_in_Forms
- https://www.w3.org/WAI/GL/wiki/Using_Aria-Invalid_to_Indicate_An_Error_Field

**Comment by bjankord** [comment]
# Tech Design
## Adding aria attributes to InputField, TextAreaField, CheckboxField, RadioField, SelectField

### InputField, TextAreaField, and SelectField

These components have a required prop named `inputId` / `selectId`.
We pass this value down to the terra-form-field component via the `htmlFor` prop.
Inside the Field component, we can use that to generate IDs for the error and help text containers.

ID for error container: ```id={`${htmlFor}-error`}```
ID for help container: ```id={`${htmlFor}-help`}```

Inside the InputField, TextAreaField, and SelectField components, we'd need to add logic to see if help text was displayed or if error text was displayed. If these are displayed, we'd add `aria-describedby` using the value from `inputId` / `selectId` and append `-help` or `-error`. These would map to the IDs set by the terra-form-field component.

#### aria-labelledby
In regards to adding `aria-labelledby` to the inputs, I don't see the reasoning behind it as the InputField, TextAreaField, and SelectField fields all require a label prop be set. In the [linked post above](https://www.w3.org/WAI/tutorials/forms/instructions/), it notes "At the time of writing this tutorial, this approach is not fully supported by all web browsers and assistive technologies [...] To ensure backward compatibility, the `for` and `id` attributes are also used in this example."

We currently use `for` and `id` in InputField, TextAreaField, and SelectField.

#### Notes
This implementation assumes users are providing unique IDs to the required prop named `inputId` or `selectId` on InputField, TextAreaField, and SelectField. I'd like to keep it this way instead of adding a random hash onto the IDs internally. I think adding a random hash to ensure IDs are unique gets into some mutation that will only add complexity. This means though that a consumer could have a form with 2 IDs that are the same. This would be on the consumer to resolve.

I'd also recommend adding a check to ensure the string value passed to the `inputId` / `selectId` does not contain whitespaces, otherwise the ID will be invalid. There is [an example currently that passes two words](https://github.com/cerner/terra-core/blob/351c95a4e629b94600fb1ca867b5f800de8e9166/packages/terra-form-input/src/terra-dev-site/doc/form-input/InputField.doc.jsx#L23) to the inputID prop.

### CheckboxField and RadioField
In CheckboxField's and RadioField's current AP design, these components are a bit more complex to implement the aria-describedby attribute on the associated inputs.

#### Add required id or fieldSetId prop
These components will need a new required prop added to them, `id` or `fieldSetId`. This would be used to set an id on the help text and error text containers. We then need to associate these IDs to the radio / checkbox inputs. We likely wouldn't use this for anything else, though we could set it as the ID on the fieldset. With this change, we should also likely make `legend` a required prop so it aligns with the API of InputField, TextareaField, and SelectField.

In regards to associating the help text container and error text container IDs to the Radio and Checkbox input, I have a few ways I've thought of that this could be done.

#### Replace children with options prop
I'd recommend we change the RadioField and CheckboxField components to take an `options` prop thats an array of options instead of using the children prop. This would allow us to bring the rendering of the radio and checkbox components into the render method of RadioField and CheckboxField and easily map the IDs for the help text and error text containers to the aria-describedby attribute on the inputs. This API would be similar to Ant Design's [RadioGroup](https://ant.design/components/radio/#components-radio-demo-radiogroup) and [Checkbox Group](https://ant.design/components/checkbox/#components-checkbox-demo-group) API. The drawback with this is it would be a non-passive change. It also limits flexibility of the rendering. 

#### Use render props with children API
If we instead keep the children API for setting Radio / Checkbox inputs, we could use a render prop to make the `aria-describedby` attribute value available to the consumer to add to the inputs.

This may look something like:

```js
<RadioField
  id="testing-radio-field"
  legend="Which Department do you work for?"
  help="Your work falls inline with one of these departments"
  error="One must be selected"
>
  {radio => (
    <React.Fragment>
      <Radio id="ux-dept" name="dept" labelText="UX/Interaction Design" onChange={this.handleOnChange} value="ux" inputAttrs={{ 'aria-describedby': radio.ariaDescribedby }} />
      <Radio id="magazine-dept" name="dept" labelText="Magazine Advertisements" onChange={this.handleOnChange} value="magazine" inputAttrs={{ 'aria-describedby': radio.ariaDescribedby }} />
      <Radio id="website-dept" name="dept" labelText="Website Advertisements" onChange={this.handleOnChange} value="website" inputAttrs={{ 'aria-describedby': radio.ariaDescribedby }} />
    </React.Fragment>
  )}
</RadioField>
```

The drawback with this is that it requires consumers to be aware of the object being passed to the children and an understanding of how to apply it to inputAttrs prop to make the mapping complete. I'd prefer we handle wiring up all the aria attributes within an encapsulated RadioField / CheckboxField component. With this approach, it feels like an implementation detail gets leaked out to the consumer for this mapping to work correctly / accessibly.

#### Compound component / clone children
With our current API of `<Radio />` components nested inside of `<RadioField />`, we could clone the children, check to see that they are `<Radio />` components, and add the `aria-describedby` to the `inputAttrs` prop on the cloned items.

```jsx
{React.Children.map(children, child => {
  if (child.type === Radio) {
    return React.cloneElement(child, { 'inputAttrs': {'aria-describedby': 'cat dog'} });
  }
  return child;
})}
 ```

This could work well and wouldn't require an API change. The drawback I see with this that it only works if the `<Radio>` components are direct children of `<RadioField>`. Maybe this constraint isn't that bad, but its something to be aware of with this approach. Its a limit of the flexibility of the rendering, but offers more flexibility compared to the the options prop idea.

 #### Using context
Using context would allow us to set the `aria-describedby` on the `<Radio />` component easily via `<RadioField>`, but it  imposes a pairing where `<Radio>` could only be rendered inside the context provider, `<RadioField>`. Since `<Radio>` is an atomic component an intended to be used both within `<RadioField>` and outside of `<RadioField>` for cases where users need to make their own radio inputs, I don't see this as a good option.

**Comment by bjankord** [comment]
Ran into this issue testing with VoiceOver on iOS.

**Comment by kschuste** [comment]
Is this issue still in discussion on getting worked on?  There is an accessibility error around aria-label for the input field that is occurring that always needs to be turned off for wdio tests.

**Comment by bjankord** [comment]
It is slated to be worked on in Q1. We'll likely start development in March.

**Comment by bjankord** [comment]
I'm splitting this issue into 2 as some of the work can be done as passive changes now and some of the work will require a major version bump. Logged https://github.com/cerner/terra-core/issues/2267 for work that should be passive to add.

**Comment by bjankord** [comment]
I've found a way to keep this passive. Including the error and help text inside the legend as visually hidden text and then mapping `aria-describedby` on the inputs to a unique ID on the legend enables the legend, error text, and help text to be read by JAWS and VoiceOver as the user interacts with the checkbox and radio buttons in CheckboxField and RadioField.
